qiskit/backends/ibmq/ibmqbackend.py

```diff
@@ -103,20 +117,38 @@ def run(self, q_job):
         if backend_name != self.name:
             raise QISKitError("inconsistent qobj backend "
                               "name ({0} != {1})".format(backend_name, self.name))
-        output = self._api.run_job(api_jobs, backend_name,
-                                   shots=qobj['config']['shots'],
-                                   max_credits=qobj['config']['max_credits'],
-                                   seed=seed0,
-                                   hpc=hpc)
-        if 'error' in output:
-            raise ResultError(output['error'])
+        submit_info = self._api.run_job(api_jobs, backend_name,
+                                        shots=qobj['config']['shots'],
+                                        max_credits=qobj['config']['max_credits'],
+                                        seed=seed0,
+                                        hpc=hpc)
+        self._submit_info = submit_info
+        return submit_info
+
+    def _run_job(self, q_job):
```

Comment:
Related to the handling of attributes and the decoupling as a whole, would moving some methods to `IBMQJob` help reducing the complexity? Currently, the flow is similar to (warning: ugly ascii graph ahead):

```
compiler        backend            job                   api
**********      ***************    ******************    ****************

.execute() ---> .run(q_job)
                    |
                ._submit(q_job) -----------------------> .run_job()
                    |
(job returned) <-- IBMQJob()
                    | (async)
                    -------------> _executor.submit()
                                          |
                ._run_job(q_job) <---------
                    |
                ._wait_for_job(job_id) ----------------> .get_job(job_id)
```

in particular, when a `IBMQJob()` instance is created, the future receives a callable _from `IBMQBackend`_, which also uses internally another function from the same object. Perhaps moving `_run_job` and `_wait_for_job` to the Job itself would keep things less complex and help with the issue of sharing variables as well?

Reply:
I guess this depends on where we want the processing to happen. Currently all backends have a `_run_job` method. This was the previously blocking call `run` in all the backends. So `_run_job` is responsible for running all the circuits in the qobj. For local backends all the "work" happens in the backends. I think it is nice to keep the provider and job classes lightweight so most of the implementation can happen in the backend. The provider and job classes can almost be directly copied to new backend chains with little change.

Reply:
moved